tests/unit/optimizers/test_lr_scheduler.py

```diff
@@ -0,0 +1,367 @@
+import torch
+import torch.nn.functional as F
+from torch.optim import SGD
+from torch.optim.lr_scheduler import _LRScheduler
+
+import math
+import numpy as np
+from tests import reset_seed
+import pytorch_lightning as pl
+from pl_bolts.optimizers.lr_scheduler import LinearWarmupCosineAnnealingLR
+
+
+EPSILON = 1e-12
+
+
+class SchedulerTestNet(torch.nn.Module):
+    """
+    adapted from: https://github.com/pytorch/pytorch/blob/master/test/test_optim.py
+    """
+    def __init__(self):
+        super(SchedulerTestNet, self).__init__()
+        self.conv1 = torch.nn.Conv2d(1, 1, 1)
+        self.conv2 = torch.nn.Conv2d(1, 1, 1)
+
+    def forward(self, x):
+        return self.conv2(F.relu(self.conv1(x)))
+
+
+class TestLRScheduler(object):
+    """
+    adapted from: https://github.com/pytorch/pytorch/blob/master/test/test_optim.py
+    """
+    def __init__(self, base_lr=0.05, multiplier=10):
+        self.base_lr = base_lr
+        self.multiplier = multiplier
+
+        self.net = SchedulerTestNet()
+        self.optimizer = SGD(
+            [
+                {'params': self.net.conv1.parameters()},
+                {'params': self.net.conv2.parameters(), 'lr': base_lr * multiplier}
+            ], lr=base_lr)
+
+        self.closed_form_net = SchedulerTestNet()
+        self.closed_form_opt = SGD(
+            [
+                {'params': self.closed_form_net.conv1.parameters()},
+                {'params': self.closed_form_net.conv2.parameters(), 'lr': base_lr * multiplier}
+            ], lr=base_lr)
+
+    def _test_lr(self, schedulers, targets, epochs=10):
+        if isinstance(schedulers, _LRScheduler):
+            schedulers = [schedulers]
+        for epoch in range(epochs):
+            for param_group, target in zip(self.optimizer.param_groups, targets):
+                assert abs(target[epoch] - param_group['lr']) < EPSILON,\
+                    "LR is wrong in epoch {}: expected {}, got {}".format(
+                        epoch, target[epoch], param_group['lr']
+                    )
+            [scheduler.step() for scheduler in schedulers]
+
+    def _test_against_closed_form(self, scheduler, closed_form_scheduler, epochs=10):
+        targets = []
+        for epoch in range(epochs):
+            closed_form_scheduler.step(epoch)
```

Comment:
this comes from earlier versions of pytorch where scheduler.step() was called before the epoch instead of after.

Reply:
but do we still support these old versions?

Reply:
@justusschock current version of pytorch still defines ``` _test_against_closed_form()``` for now, all we need to do is define that function to support the old version. I have a warning and doc available for this